File: examples/microgrid/simulators/heatpump.py
# -*- coding: utf-8 -*-
"""Air-source heat pump — external simulator for energysim.

Algebraic Carnot-fraction COP model: no internal dynamics.
Rated electrical input 3 kW.
"""

from energysim.base import SimulatorAdapter


class external_simulator(SimulatorAdapter):
    """Algebraic COP model of an air-source heat pump (no dynamics)."""

    def __init__(self, inputs=None, outputs=None, **kwargs):
        self.inputs = inputs or []
        self.outputs = outputs or []

        # I/O variables
        self.P_cmd = 0.0        # MW electrical command
        self.T_source = 5.0     # °C ambient air
        self.T_sink = 50.0      # °C storage temperature

        # Outputs
        self.Q_thermal = 0.0    # MW thermal
        self.COP = 3.0
        self.P_elec = 0.0       # MW

        # Rating
        self.max_elec = 0.003   # MW (3 kW)

    def init(self):
        pass

    def set_value(self, parameters, values):
        for p, v in zip(parameters, values):
            if p == 'P_cmd':
                self.P_cmd = v
            elif p == 'T_source':
                self.T_source = v
            elif p == 'T_sink':
                self.T_sink = v

    def get_value(self, parameters, time):
        result = []
        for p in parameters:
            if p == 'Q_thermal':
                result.append(self.Q_thermal)
            elif p == 'COP':
                result.append(self.COP)
            elif p == 'P_elec':
                result.append(self.P_elec)
            else:
                result.append(0.0)
        return result

    def step(self, time):
        dT = self.T_sink - self.T_source
        dT_safe = max(1.0, dT)

        # Carnot-fraction COP, clipped to [1, 6]
        COP = 0.45 * (self.T_sink + 273.15) / dT_safe
        self.COP = max(1.0, min(6.0, COP))

        # Clip electrical power to rating
        self.P_elec = max(0.0, min(self.max_elec, self.P_cmd))

        # Thermal output
        self.Q_thermal = self.COP * self.P_elec

    def cleanup(self):
        pass

File: examples/matlab_cosim/models/heatpump.m
function [Q_thermal, COP] = heatpump(P_electric, T_source, T_sink, time)
% HEATPUMP  Carnot-limited heat-pump model for co-simulation.
%
%   [Q_thermal, COP] = heatpump(P_electric, T_source, T_sink, time)
%
%   Inputs
%       P_electric  - Electrical power command [kW]  (>= 0)
%       T_source    - Source (outdoor / ground) temperature [degC]
%       T_sink      - Sink (indoor / water) temperature  [degC]
%       time        - Current simulation time [s]  (appended by energysim)
%
%   Outputs
%       Q_thermal   - Thermal power delivered [kW]
%       COP         - Instantaneous coefficient of performance [-]
%
%   The COP is calculated from the reversed Carnot cycle with an
%   exergetic efficiency (eta_ex) of 0.45, which is typical for
%   residential air-source heat pumps.
%
%   Example
%       [Q, cop] = heatpump(3.0, 5, 35, 0);

    % --- Parameters ---
    eta_ex     = 0.45;      % exergetic efficiency  [-]
    P_standby  = 0.05;      % standby consumption   [kW]
    COP_min    = 1.0;       % minimum COP clamp     [-]
    COP_max    = 8.0;       % maximum COP clamp     [-]

    % --- Guard: no heating requested ---
    if P_electric <= P_standby
        Q_thermal = 0.0;
        COP       = 0.0;
        return;
    end

    % --- Carnot COP ---
    T_source_K = T_source + 273.15;
    T_sink_K   = T_sink   + 273.15;
    dT         = T_sink_K - T_source_K;

    if dT <= 0
        % Source is warmer than sink – free cooling, COP is very high
        COP_carnot = COP_max;
    else
        COP_carnot = T_sink_K / dT;
    end

    COP = eta_ex * COP_carnot;
    COP = max(COP_min, min(COP, COP_max));   % clamp

    % --- Thermal output ---
    Q_thermal = P_electric * COP;
end
